### PR TITLE
replace sequential op with the new type Seq in DAG and AST

### PR DESCRIPTION
**Pull request by mccanne**
This commit replaces the sequential operator with a slice of ops called Seq imposing a more consistent and robust pattern on the DAG.  This prevents us from playing games with sneaking sequentials inside of sequentials as this was a questionable practice that made downstream logic error prone.

This should make the DAG more robustness and easier to evolve as we add support for vector ops.

To accommodate this approach, we added ast.Sample for the sample op and put the logic to expand it into two ops in the semantic pass where this belongs.

We also separated dag.Parallel into dag.Fork and dag.Scatter to get rid of the Any flag inside of Parallel.

### CODE REVIEW

File: compiler/optimizer/optimizer.go [hunk 1]
```diff
@@ -104,172 +130,164 @@ func walkOp(op dag.Op, over bool, post func(dag.Op)) {
...
+		o.propagateSortKey(seq, order.Nil)
+		// See if we can lift a filtering predicate into the source op.
+		// Filter might be nil in which case we just put the chain back
+		// on the source op and zero out the source's filter.
+		filter, chain := matchFilter(chain)
+		switch op := seq[0].(type) {
+		case *dag.PoolScan:
+			// Here we transform a PoolScan into a Lister followed by one or more chains
+			// of slicers and sequence scanners.  We'll eventually choose other configurations
+			// here based on metadata and availability of vzng.
```
Comment (nwt) [inline compiler/optimizer/optimizer.go:211]:
```suggestion
			// here based on metadata and availability of VNG.
```

File: compiler/optimizer/optimizer.go [hunk 2]
```diff
@@ -104,172 +130,164 @@ func walkOp(op dag.Op, over bool, post func(dag.Op)) {
...
+			lister := &dag.Lister{
+				Kind:   "Lister",
+				Pool:   op.ID,
+				Commit: op.Commit,
+			}
+			// Check to see if we can add a range pruner when the pool-key is used
```
Comment (nwt) [inline compiler/optimizer/optimizer.go:217]:
```suggestion
			// Check to see if we can add a range pruner when the pool key is used
```

File: compiler/semantic/op.go [hunk 3]
```diff
@@ -489,31 +469,31 @@ func semOp(ctx context.Context, scope *Scope, o ast.Op, ds *data.Source, head *l
...
 	case *ast.Uniq:
-		return &dag.Uniq{
+		return append(seq, &dag.Uniq{
 			Kind:  "Uniq",
 			Cflag: o.Cflag,
-		}, nil
+		}), nil
 	case *ast.Pass:
-		return &dag.Pass{Kind: "Pass"}, nil
+		return append(seq, &dag.Pass{Kind: "Pass"}), nil
```
Comment (nwt) [inline compiler/semantic/op.go:482]:
```suggestion
		return append(seq, dag.PassOp), nil
```

File: compiler/optimizer/optimizer.go [hunk 4]
```diff
@@ -104,172 +130,164 @@ func walkOp(op dag.Op, over bool, post func(dag.Op)) {
...
+		case *dag.FileScan:
+			op.Filter = filter
+			seq = append(dag.Seq{op}, chain...)
+		case *dag.CommitMetaScan:
+			if op.Tap {
+				sortKey, err := o.sortKeyOfSource(op)
+				if err != nil {
+					return nil, err
+				}
+				// Check to see if we can add a range pruner when the pool-key is used
```
Comment (nwt) [inline compiler/optimizer/optimizer.go:248]:
```suggestion
				// Check to see if we can add a range pruner when the pool key is used
```

File: compiler/optimizer/optimizer.go [hunk 5]
```diff
@@ -104,172 +130,164 @@ func walkOp(op dag.Op, over bool, post func(dag.Op)) {
...
+				// in a normal filtering operation.
+				if !sortKey.IsNil() && filter != nil {
+					if p := newRangePruner(filter, sortKey.Primary(), sortKey.Order); p != nil {
+						op.KeyPruner = p
+					}
+				}
```
Comment (nwt) [inline compiler/optimizer/optimizer.go:254]:
```suggestion
				op.KeyPruner = maybeNewRangePruner(filter, sortKey)
```

File: compiler/optimizer/optimizer.go [hunk 6]
```diff
@@ -104,172 +130,164 @@ func walkOp(op dag.Op, over bool, post func(dag.Op)) {
...
+				// Delete the downstream operators when we are tapping the object list.
+				seq = dag.Seq{op}
+			}
+		case *kernel.Reader:
+			op.Filter = filter
+			seq = append(dag.Seq{op}, chain...)
+		case *dag.LakeMetaScan, *dag.PoolMetaScan, *dag.HTTPScan:
+		default:
+			return nil, fmt.Errorf("an entry point to the query is not a source: %T", op)
```
Comment (nwt) [inline compiler/optimizer/optimizer.go:263]:
```suggestion
			return nil, fmt.Errorf("internal error: an entry point is not a source: %T", op)
```

File: compiler/optimizer/optimizer.go [hunk 7]
```diff
@@ -104,172 +130,164 @@ func walkOp(op dag.Op, over bool, post func(dag.Op)) {
...
-	case *dag.LakeMetaScan, *dag.PoolMetaScan, *dag.HTTPScan:
-	default:
-		return fmt.Errorf("internal error:  point to the query is not a source: %T", op)
-	}
-	return nil
+		return seq, nil
+	})
 }
 
+// XXX update comment
```
Comment (nwt) [inline compiler/optimizer/optimizer.go:269]: Please do.

File: compiler/optimizer/parallelize.go [hunk 8]
```diff
@@ -23,14 +22,14 @@ func (o *Optimizer) parallelizeScan(ops []dag.Op, replicas int) ([]dag.Op, error
...
-	scan, ok := ops[0].(*dag.SeqScan)
-	if !ok {
-		return nil, errors.New("internal error: parallelizeScan: entry has no dag.SeqScan")
-	}
-	if len(ops) == 1 && scan.Filter == nil {
+	return nil, errors.New("parallelization of non-pool queries is not yet supported")
+}
+
+func (o *Optimizer) parallelizeSeqScan(scan *dag.SeqScan, ops []dag.Op, replicas int) ([]dag.Op, error) {
+	if len(ops) == 1 && scan.Filter == nil { //XXX
```
Comment (nwt) [inline compiler/optimizer/parallelize.go:32]: Nit: Add some detail to this XXX comment.

File: compiler/optimizer/parallelize.go [hunk 9]
```diff
@@ -119,11 +121,10 @@ func (o *Optimizer) liftIntoParPaths(ops []dag.Op) {
 			// Need an unmodified summarize to split into its parials pieces.
 			return
 		}
-		for _, o := range par.Ops {
-			path := o.(*dag.Sequential)
-			partial := copyOp(op).(*dag.Summarize)
+		for k := 0; k < len(fork.Paths); k++ {
```
Comment (nwt) [inline compiler/optimizer/parallelize.go:124]:
```suggestion
		for k :=  range fork.Paths {
```

File: compiler/optimizer/parallelize.go [hunk 10]
```diff
@@ -149,10 +150,9 @@ func (o *Optimizer) liftIntoParPaths(ops []dag.Op) {
 				return
 			}
 		}
-		for _, o := range par.Ops {
-			path := o.(*dag.Sequential)
-			sort := copyOp(op).(*dag.Sort)
-			path.Ops = append(path.Ops, sort)
+		for k := 0; k < len(fork.Paths); k++ {
+			sort := CopyOp(op).(*dag.Sort)
+			fork.Paths[k].Append(sort)
```
Comment (nwt) [inline compiler/optimizer/parallelize.go:155]:
```suggestion
		for k := range fork.Paths {
			fork.Paths[k].Append(copyOp(op))
```

File: compiler/optimizer/parallelize.go [hunk 11]
```diff
@@ -174,51 +174,32 @@ func (o *Optimizer) liftIntoParPaths(ops []dag.Op) {
 	case *dag.Head, *dag.Tail:
 		// Copy the head or tail into the parallel path and leave the original in
 		// place which will apply another head or tail after the merge.
-		for _, o := range par.Ops {
-			path := o.(*dag.Sequential)
-			path.Ops = append(path.Ops, copyOp(op))
+		for k := 0; k < len(fork.Paths); k++ {
```
Comment (nwt) [inline compiler/optimizer/parallelize.go:177]:
```suggestion
		for k := range ork.Paths {
```

File: compiler/optimizer/parallelize.go [hunk 12]
```diff
@@ -174,51 +174,32 @@ func (o *Optimizer) liftIntoParPaths(ops []dag.Op) {
...
 			if err != nil || !key.Equal(mergeKey) {
 				// This operator destroys the merge order so we cannot
 				// lift it up into the parallel legs in front of the merge.
 				return
 			}
 		}
-		for _, o := range par.Ops {
-			path := o.(*dag.Sequential)
-			path.Ops = append(path.Ops, copyOp(op))
+		for k := 0; k < len(fork.Paths); k++ {
```
Comment (nwt) [inline compiler/optimizer/parallelize.go:195]:
```suggestion
		for k := range fork.Paths {
```

File: compiler/parser/parser.peg [hunk 13]
```diff
@@ -120,19 +118,19 @@ Operation
 EndOfOp = __ (Pipe / SearchKeywordGuard / "=>" / ")" / EOF)
 Pipe = "|" !("{" / "[")
 
-Leg = __ "=>" __ seq:Sequential { RETURN(seq) }
+Leg = __ "=>" __ seq:Seq { RETURN(seq) }
 
 SwitchLeg
   = __ expr:Case leg:Leg {
-      RETURN(MAP("expr": expr, "op": leg))
+      RETURN(MAP("expr": expr, "path": leg))
```
Comment (nwt) [inline compiler/parser/parser.peg:125]: Nit: s/leg/path/

File: compiler/semantic/op.go [hunk 14]
```diff
@@ -255,26 +259,18 @@ func semPoolWithName(ctx context.Context, scope *Scope, p *ast.Pool, poolName st
...
+			ID:     poolID,
+			Commit: commitID,
 		}, nil
 	}
 	return &dag.PoolScan{
 		Kind:   "PoolScan",
 		ID:     poolID,
 		Commit: commitID,
 	}, nil
+
```
Comment (nwt) [inline compiler/semantic/op.go:273]:
```suggestion
```

File: compiler/kernel/expr.go [hunk 15]
```diff
@@ -473,8 +473,8 @@ func (b *Builder) compileMapExpr(m *dag.MapExpr) (expr.Evaluator, error) {
 }
 
 func (b *Builder) compileOverExpr(over *dag.OverExpr) (expr.Evaluator, error) {
-	if over.Scope == nil {
-		return nil, errors.New("over expression requires flow body")
+	if over.Body == nil {
+		return nil, errors.New("over expression requires a query body")
```
[the remark this answers is not in the document]
Comment (mccanne) [reply compiler/kernel/expr.go:477]: I think "lateral query body" is more informative since that's what the docs call it and this is user facing.

**Review by nwt — APPROVED** [review]

**Pull request merged**